pkg/trafficrouting/network/composite.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2024 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package network
+
+import (
+	"context"
+
+	"go.uber.org/multierr"
+
+	"github.com/openkruise/rollouts/api/v1beta1"
+)
+
+var (
+	_ NetworkProvider = (CompositeController)(nil)
+)
+
+// CompositeController is a set of NetworkProvider
+type CompositeController []NetworkProvider
+
+func (c CompositeController) Initialize(ctx context.Context) error {
+	var err error
+	for _, provider := range c {
+		err = multierr.Append(err, provider.Initialize(ctx))
+	}
+	return err
+}
+
+func (c CompositeController) EnsureRoutes(ctx context.Context, strategy *v1beta1.TrafficRoutingStrategy) (bool, error) {
```

Comment:
I suppose the only point in dispute is how to handle error from a single networkProvider.

Can you have a check? @zmberg @furykerry